src/data/set/pointwise.lean

```diff
@@ -261,60 +261,79 @@ localized "attribute [instance] set.mul_one_class set.add_zero_class set.semigro
+
+section monoid
+variables [monoid α]
+
+/-- If `u` is a unit in `α`, then `{u}` is a unit in `set α`. -/
+@[to_additive "If `u` is an additive unit in `α`, then `{u}` is an additive unit in `set α`.",
+simps coe] protected def _root_.units.set (u : αˣ) : (set α)ˣ :=
+{ val := {u},
+  inv := {(u⁻¹ : αˣ)},
+  val_inv := by rw [singleton_mul_singleton, units.mul_inv, singleton_one],
+  inv_val := by rw [singleton_mul_singleton, units.inv_mul, singleton_one] }
+
+@[simp, to_additive] lemma _root_.is_unit.inv_set (u : αˣ) : u.set⁻¹ = u⁻¹.set := rfl
+
+@[to_additive] protected lemma _root_.is_unit.set (ha : is_unit a) : is_unit ({a} : set α) :=
+by { obtain ⟨u, rfl⟩ := ha, exact ⟨u.set, rfl⟩ }
+
+@[to_additive] lemma pow_mem_pow (ha : a ∈ s) : ∀ n : ℕ, a ^ n ∈ s ^ n
+| 0 := by { rw pow_zero, exact one_mem_one }
+| (n + 1) := by { rw pow_succ, exact mul_mem_mul ha (pow_mem_pow _) }
+
+@[to_additive] lemma pow_subset_pow (hst : s ⊆ t) : ∀ n : ℕ, s ^ n ⊆ t ^ n
+| 0 := by { rw pow_zero, exact subset.rfl }
+| (n + 1) := by { rw pow_succ, exact mul_subset_mul hst (pow_subset_pow _) }
+
+@[to_additive] lemma empty_pow (n : ℕ) (hn : n ≠ 0) : (∅ : set α) ^ n = ∅ :=
 by rw [← tsub_add_cancel_of_le (nat.succ_le_of_lt $ nat.pos_of_ne_zero hn), pow_succ, empty_mul]
 
-instance decidable_mem_mul [monoid α] [fintype α] [decidable_eq α]
-  [decidable_pred (∈ s)] [decidable_pred (∈ t)] :
+-- can be generalized to `[fintype s] [fintype t]`
+@[to_additive]
+instance decidable_mem_mul [fintype α] [decidable_eq α] [decidable_pred (∈ s)]
+  [decidable_pred (∈ t)] :
   decidable_pred (∈ s * t) :=
 λ _, decidable_of_iff _ mem_mul.symm
 
-instance decidable_mem_pow [monoid α] [fintype α] [decidable_eq α]
-  [decidable_pred (∈ s)] (n : ℕ) :
+@[to_additive]
+instance decidable_mem_pow [fintype α] [decidable_eq α] [decidable_pred (∈ s)] (n : ℕ) :
   decidable_pred (∈ (s ^ n)) :=
 begin
   induction n with n ih,
   { simp_rw [pow_zero, mem_one], apply_instance },
   { letI := ih, rw pow_succ, apply_instance }
 end
 
-@[to_additive]
-lemma subset_mul_left [mul_one_class α] (s : set α) {t : set α} (ht : (1 : α) ∈ t) : s ⊆ s * t :=
-λ x hx, ⟨x, 1, hx, ht, mul_one _⟩
-
-@[to_additive]
-lemma subset_mul_right [mul_one_class α] {s : set α} (t : set α) (hs : (1 : α) ∈ s) : t ⊆ s * t :=
-λ x hx, ⟨1, x, hs, hx, one_mul _⟩
-
-lemma pow_subset_pow [monoid α] (hst : s ⊆ t) (n : ℕ) :
-  s ^ n ⊆ t ^ n :=
-begin
-  induction n with n ih,
-  { rw pow_zero,
-    exact subset.rfl },
-  { rw [pow_succ, pow_succ],
-    exact mul_subset_mul hst ih },
-end
-
-@[simp, to_additive]
-lemma univ_mul_univ [monoid α] : (univ : set α) * univ = univ :=
+@[simp, to_additive] lemma univ_mul_univ : (univ : set α) * univ = univ :=
 begin
-  have : ∀x, ∃a b : α, a * b = x := λx, ⟨x, ⟨1, mul_one x⟩⟩,
+  have : ∀ x, ∃ a b : α, a * b = x := λ x, ⟨x, 1, mul_one x⟩,
   simpa only [mem_mul, eq_univ_iff_forall, mem_univ, true_and]
 end
 
+--TODO: `to_additive` trips up on the `1 : ℕ` used in the pattern-matching.
```

Comment:
You should still call `to_additive` manually afterwards so that any lemma using these can still work